src/uct/tcp/tcp_ep.c

```diff
@@ -1611,50 +1662,56 @@ ucs_status_t uct_tcp_ep_am_short(uct_ep_h uct_ep, uint8_t am_id, uint64_t header
-    hdr->length = payload_length = length + sizeof(header);
+    hdr->length = payload_length;
 
     if (length <= iface->config.sendv_thresh) {
         uct_am_short_fill_data(hdr + 1, header, payload, length);
         status = uct_tcp_ep_am_send(ep, hdr);
-        if (ucs_unlikely(status != UCS_OK)) {
-            return status;
-        }
-
-        UCT_TL_EP_STAT_OP(&ep->super, AM, SHORT, payload_length);
     } else {
-        iov[0].iov_base = hdr;
-        iov[0].iov_len  = sizeof(*hdr);
+        status = uct_tcp_ep_am_short_sendv(ep, iface, hdr, header, payload, length,
+                                           iov, 0);
+    }
 
-        iov[1].iov_base = &header;
-        iov[1].iov_len  = sizeof(header);
+    if (ucs_unlikely(status != UCS_OK)) {
+        return status;
+    }
 
-        iov[2].iov_base = (void*)payload;
-        iov[2].iov_len  = length;
+    UCT_TL_EP_STAT_OP(&ep->super, AM, SHORT, payload_length);
+    return status;
+}
 
-        status = uct_tcp_ep_am_sendv(ep, 1, hdr, iface->config.tx_seg_size,
-                                     &header, iov, UCT_TCP_EP_AM_SHORTV_IOV_COUNT);
-        if (ucs_unlikely(status != UCS_OK)) {
-            return status;
-        }
+ucs_status_t uct_tcp_ep_am_short_iov(uct_ep_h uct_ep, uint8_t am_id,
+                                     const uct_iov_t *uct_iov, size_t uct_iov_cnt)
+{
+    uct_tcp_ep_t *ep        = ucs_derived_of(uct_ep, uct_tcp_ep_t);
+    uct_tcp_iface_t *iface  = ucs_derived_of(uct_ep->iface, uct_tcp_iface_t);
+    uct_tcp_am_hdr_t *hdr   = NULL;
+    struct iovec *iov       = ucs_alloca((uct_iov_cnt + 1) * sizeof(*iov));
+    uint32_t payload_length = uct_iov_total_length(uct_iov, uct_iov_cnt);
+    ucs_status_t status;
 
-        UCT_TL_EP_STAT_OP(&ep->super, AM, SHORT, payload_length);
+    UCT_CHECK_AM_ID(am_id);
+    UCT_CHECK_IOV_SIZE(uct_iov_cnt, iface->config.max_iov, "am_short_iov");
+    UCT_CHECK_LENGTH(payload_length, 0,
+                     iface->config.tx_seg_size - sizeof(uct_tcp_am_hdr_t),
+                     "am_short_iov");
 
-        if (uct_tcp_ep_ctx_buf_need_progress(&ep->tx)) {
-            /* Copy only user's header and payload to the TX buffer,
-             * TCP AM header is placed at the beginning of the buffer */
-            offset = ((ep->tx.offset >= sizeof(*hdr)) ?
-                      (ep->tx.offset - sizeof(*hdr)) : 0);
+    status = uct_tcp_ep_am_prepare(iface, ep, am_id, &hdr);
+    if (status != UCS_OK) {
+        return status;
+    }
 
-            ucs_iov_copy(&iov[1], UCT_TCP_EP_AM_SHORTV_IOV_COUNT - 1,
-                         offset, UCS_PTR_BYTE_OFFSET(hdr + 1, offset),
-                         ep->tx.length - sizeof(*hdr) - offset,
-                         UCS_IOV_COPY_TO_BUF);
-        }
+    ucs_assertv(hdr != NULL, "ep=%p", ep);
+    hdr->length = payload_length;
+
+    status = uct_tcp_ep_am_short_sendv(ep, iface, hdr, 0, uct_iov, uct_iov_cnt, iov,
+                                       1);
```

Comment:
can combine with the previous line?

Reply:
The line exceeds 85 characters in this case.

Reply:
> The line exceeds 85 characters in this case.

ok, I see